### PR TITLE
test: ensure stability

### LINKED ISSUE #44: Ensure stability when possible.
During the implementation of `on` upstream I noticed that the listeners were being setup and destroyed **a ton**. This is because the modifiers `updateModifier` hook is called whenever the arguments may have been changed, which in the case of `{{on 'some-string' this.whatever}}` means that any time _`this`_ is updated we rerender. In the upstream implementation I do some extra backbending to ensure that we only call `removeEventListener`/`addEventListener` in updateModifier when absolutely required.

I think the following is a rough checklist to accomplish the stability we want for this as a polyfill:

* Add a small snippet to track counts of add/remove

```js
let adds = 0;
let removes = 0;
export function counts() { 
  return { adds, removes };
}

function setupListener(...) {
  adds++;
  // ...snip...
}

function destroyListener(...) {
  removes++;
  // ...snip...
}
```

* Add a helpful `assert.counts` to the assert object:

```js
// tests/test-helper.js
import { counts } from 'ember-on-modifier/modifiers/on';

QUnit.testStart(function() {
  this._startingCounts = counts();
});

QUnit.assert.counts = function(expected) {
  let current = counts();

  this.assert.deepEqual(
    current,
    {
      adds: expected.adds + this._startingCounts.adds,
      removes: expected.removes + this._startingCounts.removes,
    },
    `counters have incremented by ${JSON.stringify(expected)}`
  );
};
```

* Add some `assert.counts({ adds: 1, removes: 0})` style assertions to tests (will likely fail for "simple" things)
* Add in manual stability checks in `updateModifier`, specifically checking if the named arguments have _actually_ changed (e.g. doing `oldCallback !== newCallback` guards)

### LINKED ISSUE #46: Refactor tests to use `assert.verifySteps` API
Totally not an issue, but I wanted to mention that I've started using `assert.step` / `assert.verifySteps` for these sorts of "order tracking" use cases. This test would look like:

```js
  test('{{on "click" (stop-propagation) capture=true}}', async function(assert) {
    assert.expect(1);

    this.outerListener = () => assert.step('outer');
    this.innerListener = () => assert.step('inner');

     await render(hbs`
      <div {{on "click" (stop-propagation this.outerListener) capture=true}}>
        <button {{on "click" this.innerListener}}>inner</button>
      </div>
    `);

     await click('button');

     assert.verifySteps(['outer'], 'it only runs the outer listener');
  });
```

See:

* https://api.qunitjs.com/assert/step
* https://api.qunitjs.com/assert/verifySteps

_Originally posted by @rwjblue in https://github.com/buschtoens/ember-on-modifier/pull/43_

### PR DESCRIPTION
**Pull request by buschtoens**
Closes #44. Also closes #46.

Apparently we were doing it right already. 🤷‍♂️ 😄

### CODE REVIEW

File: tests/test-helper.js [hunk 1]
```diff
@@ -2,6 +2,29 @@ import Application from '../app';
 import config from '../config/environment';
 import { setApplication } from '@ember/test-helpers';
 import { start } from 'ember-qunit';
+import QUnit from 'qunit';
+import { __counts } from 'ember-on-modifier/modifiers/on';
+
+let startingCounts;
+QUnit.testStart(() => {
+  startingCounts = __counts();
```
Comment (buschtoens) [inline tests/test-helper.js]: `testStart` has no `this` context, sadly. The alternative would've been a `setupAssertCounts(hooks)` test helper that adds an `assertCount` util to `this`, but I think this is still a bit cleaners.

What do you think?

Comment (rwjblue) [reply tests/test-helper.js]:
```suggestion
  QUnit.config.current.testEnvironment._startingCounts = __counts();
```

`QUnit.config.current.testEnvironment ` represents the `this` context of the test

File: tests/test-helper.js [hunk 2]
```diff
@@ -2,6 +2,29 @@ import Application from '../app';
...
+QUnit.assert.counts = function(
+  expected,
+  msg = `counters have incremented by ${JSON.stringify(expected)}`
+) {
+  const current = __counts();
+
+  this.deepEqual(
+    current,
+    {
+      adds: expected.adds + startingCounts.adds,
```
Comment (rwjblue) [inline tests/test-helper.js]:
```suggestion
      adds: expected.adds + this._startingCounts.adds,
```

File: tests/test-helper.js [hunk 3]
```diff
@@ -2,6 +2,29 @@ import Application from '../app';
...
+      removes: expected.removes + startingCounts.removes
```
Comment (rwjblue) [inline tests/test-helper.js]:
```suggestion
      removes: expected.removes + this._startingCounts.removes
```

**Comment by rwjblue** [comment]
> Apparently we were doing it right already. 🤷‍♂️ 😄

I'm surprised by this a bit, but the tests look great and thank you very much for doing the assertions!

**Comment by buschtoens** [comment]
I was also surprised. 😅

I guess the Ember public Helper API diffs the arguments, whereas the private Glimmer API does not.

**Pull request merged**